Auto debug signals from list

lekcyjna123 opened the pull request:
As a part of debuging #260 I have found that `auto_debug_signals` have problems when methods are in list. Here is extension which add support for this case.

tilk commented on coreblocks/utils/debug_signals.py:
```diff
@@ -27,6 +30,17 @@ def auto_debug_signals(thing) -> SignalBundle:
             smap[v] = a
         elif isinstance(a, Signal) or isinstance(a, Record):
             slist.append(a)
+        elif isinstance(a, list):
```
Is generalizing this to `Collection` useful?

lekcyjna123 replied:
Yes. In case of list of methods which can have length depending on generation parameters. E.g. in multiport fifo number of ports depends on parameters given by user, so methods are stored in the list.

On the other hand we can decide to don't store methods in list and instead of that use `self[f"method_name_{i}"]` but I am not sure if this don't constrain us in future.

tilk replied:
You didn't answer the question...

lekcyjna123 replied:
Sorry. I wrongly understand your question. I will try to generalize.

lekcyjna123 replied:
Fixed.

lekcyjna123 commented on coreblocks/utils/debug_signals.py:
```diff
@@ -27,6 +30,17 @@ def auto_debug_signals(thing) -> SignalBundle:
         elif isinstance(a, Signal) or isinstance(a, Record):
             slist.append(a)
+        elif isinstance(a, list):
+            submap = {}
```
Because I want to have mapping name -> signals similar as in `smap`

lekcyjna123 commented on coreblocks/utils/debug_signals.py:
```diff
@@ -27,6 +30,17 @@ def auto_debug_signals(thing) -> SignalBundle:
+        elif isinstance(a, list):
+            submap = {}
+            for i, e in enumerate(a):
+                if isinstance(e, HasDebugSignals):
+                    submap[f"{v}[{i}]"] = e.debug_signals()
+                else:
+                    sublist = auto_debug_signals(e)
+                    if sublist:
+                        submap[f"{v}[{i}]"] = sublist
```
Yes. It is some duplication of functionality, but without that generation from lists don't worked. If I call `auto_debug_signals` directly on `e`, then I will have infinity loop, because `e` which is potentially a `Method` will be treated as `Module` and there will be an iteration over all `for v in vars(thing)`

lekcyjna123 replied:
I probably found one reference cycle. In TransactionBase we store `self.owner : Elaboratable`, which is needed for graphs. So when we iterate over vars of method we see it and by recursion we call `auto_debug_signals` after few steps of processing owner we are back in this same method.

lekcyjna123 commented:
I have rewritten this function so that calling `auto_debug_signals` on element with `debug_signal` will correctly call `debug_signal` method. This also break reference cycle, because we don't traverse `owner` in instances of `TransactionBase`

Arusekk reviewed (approved):
Noice.

lekcyjna123 commented:
I have found another reference cycle. This time in Amaranth, which use MustUse class which point on it self to check if Elaboratables are used. So I introduced following changes:
- added support for Mapping
- added visited set to break dependecy cycles
- added ignoring of private fields to get pretty print (to don't print _MustUse_context)

I acknowledge that last point can be controversial, so I am ready to remove it

Arusekk commented on coreblocks/utils/debug_signals.py:
```diff
@@ -1,8 +1,10 @@
 
 
-def auto_debug_signals(thing) -> SignalBundle:
+def auto_debug_signals(thing, *, visited: set = set()) -> SignalBundle:
```
Note this is a potential footgun, because this function is not reentrant (the set is instantiated once on function definition).

lekcyjna123 replied:
Added docstring.

Arusekk commented on coreblocks/utils/debug_signals.py:
```diff
@@ -14,19 +16,51 @@ def auto_debug_signals(thing) -> SignalBundle:
+            continue
+        visited.add(id(a))
+
+        # ignore private fields (mostly to ignore _MustUse_context to get preaty print)
```
```suggestion
        # ignore private fields (mostly to ignore _MustUse_context to get pretty print)
```

tilk commented on coreblocks/utils/debug_signals.py:
```diff
@@ -1,32 +1,71 @@
+                    submap[f"{v}[{i}]"] = sublist
+            if submap:
+                smap[v] = submap
+        elif isinstance(a, Collection):
```
Maybe check if `a` is not a `str`? A silly special case, I know, but...

lekcyjna123 commented on coreblocks/utils/debug_signals.py:
```diff
@@ -11,22 +13,65 @@ def auto_debug_signals(thing) -> SignalBundle:
+            # ignore private fields (mostly to ignore _MustUse_context to get pretty print)
+            if v[0] == "_":
+                continue
+
```
Done

lekcyjna123 commented on coreblocks/utils/debug_signals.py:
```diff
@@ -11,22 +13,65 @@ def auto_debug_signals(thing) -> SignalBundle:
+            # string being `Collection`s
+            elif isinstance(a, str):
+                continue
+            elif isinstance(a, Collection):
```
Function has been rewritten

tilk reviewed (approved):
Let's finally merge this, hopefully it will work well.
